### PR TITLE
[auth] Thread safe auth methods with recursive mutex

### PR DESCRIPTION
**Pull request by elpaso**
Add a mutex to the base class and use that
mutex to protect all public methods of the
authentication methods.

### CODE REVIEW

File: src/core/auth/qgsauthmethod.h [hunk 1]
```diff
@@ -190,6 +196,8 @@ class CORE_EXPORT QgsAuthMethod : public QObject
     QgsAuthMethod::Expansions mExpansions;
     QStringList mDataProviders;
     int mVersion = 0;
+    QMutex *mMutex;
```
Comment (m-kuhn) [inline src/core/auth/qgsauthmethod.h]: Could this be a plain variable rather than a pointer? Would make memory management a bit safer.

Comment (elpaso) [reply src/core/auth/qgsauthmethod.h]: @m-kuhn, it was a plain variable. But I could not find a way to force the recursive behavior with the default ctor. Do you know a way to do that?

Comment (m-kuhn) [reply src/core/auth/qgsauthmethod.h]: QgsAuthMethod::QgsAuthMethod
    : mMutex( QMutex::Recursive )

Comment (elpaso) [reply src/core/auth/qgsauthmethod.h]: thanks!

Comment (m-kuhn) [reply src/core/auth/qgsauthmethod.h]: np, sorry for the shorthand style, I'm a bit in a rush today ;)

**Pull request merged**